File: src/lib/rbac.ts
import type { Role } from "./types";

// ============================================================
// Role-Based Access Control
// Source of truth: ProjectDetails.md §5 + PROBLEM_STATEMENT.md §2
// ============================================================

export type ModuleKey =
  | "dashboard"
  | "fleet"
  | "drivers"
  | "trips"
  | "maintenance"
  | "expenses"
  | "analytics"
  | "settings";

export type Access = "full" | "read" | "none";

export interface ModuleMeta {
  key: ModuleKey;
  label: string;
  icon: string; // Material Symbols name
  path: string;
}

export const MODULES: ModuleMeta[] = [
  { key: "dashboard", label: "Dashboard", icon: "dashboard", path: "/dashboard" },
  { key: "fleet", label: "Fleet", icon: "local_shipping", path: "/fleet" },
  { key: "drivers", label: "Drivers", icon: "badge", path: "/drivers" },
  { key: "trips", label: "Trips", icon: "route", path: "/trips" },
  { key: "maintenance", label: "Maintenance", icon: "build", path: "/maintenance" },
  { key: "expenses", label: "Fuel & Expenses", icon: "local_gas_station", path: "/expenses" },
  { key: "analytics", label: "Analytics", icon: "analytics", path: "/analytics" },
  { key: "settings", label: "Settings", icon: "settings", path: "/settings" },
];

// role -> module -> access level
const MATRIX: Record<Role, Record<ModuleKey, Access>> = {
  fleet_manager: {
    dashboard: "full",
    fleet: "full",
    drivers: "full",
    trips: "read",
    maintenance: "full",
    expenses: "read",
    analytics: "read",
    settings: "full",
  },
  dispatcher: {
    dashboard: "full",
    fleet: "read",
    drivers: "none",
    trips: "full",
    maintenance: "read",
    expenses: "none",
    analytics: "none",
    settings: "read",
  },
  safety_officer: {
    dashboard: "full",
    fleet: "read",
    drivers: "full",
    trips: "read",
    maintenance: "full",
    expenses: "none",
    analytics: "none",
    settings: "read",
  },
  financial_analyst: {
    dashboard: "full",
    fleet: "read",
    drivers: "none",
    trips: "none",
    maintenance: "read",
    expenses: "full",
    analytics: "full",
    settings: "read",
  },
};

export function access(role: Role, module: ModuleKey): Access {
  return MATRIX[role]?.[module] ?? "none";
}

export function canView(role: Role, module: ModuleKey): boolean {
  return access(role, module) !== "none";
}

export function canEdit(role: Role, module: ModuleKey): boolean {
  return access(role, module) === "full";
}

/** Modules visible to a role, in sidebar order. */
export function visibleModules(role: Role): ModuleMeta[] {
  return MODULES.filter((m) => canView(role, m.key));
}

/** First landing route a role is allowed to see. */
export function landingPath(role: Role): string {
  const first = visibleModules(role)[0];
  return first?.path ?? "/dashboard";
}

// Columns rendered by the Settings RBAC matrix widget
export const RBAC_MATRIX_COLUMNS: { key: ModuleKey; label: string }[] = [
  { key: "fleet", label: "Fleet" },
  { key: "drivers", label: "Drivers" },
  { key: "trips", label: "Trips" },
  { key: "expenses", label: "Fuel/Exp" },
  { key: "analytics", label: "Analytics" },
];

File: transitops/rbac.py
"""Role-Based Access Control.

One login, four roles. Access levels per module: none / view / full.
This matrix mirrors screen 8 of the wireframe:

    ROLE              FLEET  DRIVERS  TRIPS  FUEL/EXP.  ANALYTICS
    Fleet Manager       x       x       -        -          x
    Dispatcher        view      -       x        -          -
    Safety Officer      -       x     view       -          -
    Financial Analyst view      -       -        x          x
"""
import functools

from flask import g, jsonify

LEVELS = {"none": 0, "view": 1, "full": 2}

ROLES = {
    "fleet_manager": {
        "label": "Fleet Manager",
        "perms": {
            "dashboard": "view", "fleet": "full", "drivers": "full",
            "trips": "none", "maintenance": "full", "fuel": "none",
            "analytics": "view", "settings": "full",
        },
    },
    "dispatcher": {
        "label": "Dispatcher",
        "perms": {
            "dashboard": "view", "fleet": "view", "drivers": "none",
            "trips": "full", "maintenance": "none", "fuel": "none",
            "analytics": "none", "settings": "none",
        },
    },
    "safety_officer": {
        "label": "Safety Officer",
        "perms": {
            "dashboard": "view", "fleet": "none", "drivers": "full",
            "trips": "view", "maintenance": "none", "fuel": "none",
            "analytics": "none", "settings": "none",
        },
    },
    "financial_analyst": {
        "label": "Financial Analyst",
        "perms": {
            "dashboard": "view", "fleet": "view", "drivers": "none",
            "trips": "none", "maintenance": "view", "fuel": "full",
            "analytics": "full", "settings": "none",
        },
    },
}

# Columns shown in the Settings RBAC matrix (and mirrored by the frontend)
RBAC_MATRIX_COLUMNS = [
    ("fleet", "Fleet"), ("drivers", "Drivers"), ("trips", "Trips"),
    ("maintenance", "Maint."), ("fuel", "Fuel/Exp."), ("analytics", "Analytics"),
]


def role_label(role):
    return ROLES.get(role, {}).get("label", role)


def serialize_user(user):
    """Shape sent to the React app after login / on session restore."""
    role = user["role"]
    return {
        "id": user["id"],
        "name": user["name"],
        "email": user["email"],
        "role": role,
        "role_label": ROLES[role]["label"],
        "perms": ROLES[role]["perms"],
    }


def can(module, level="view"):
    """True when the logged-in user's role grants at least `level` on `module`."""
    user = g.get("user")
    if user is None:
        return False
    have = ROLES[user["role"]]["perms"].get(module, "none")
    return LEVELS[have] >= LEVELS[level]


def require(module, level="view"):
    """API decorator: 401 for anonymous callers, 403 for under-privileged roles."""
    def decorator(view):
        @functools.wraps(view)
        def wrapped(**kwargs):
            if g.get("user") is None:
                return jsonify(error="Authentication required."), 401
            if not can(module, level):
                return jsonify(error="Your role does not permit this action."), 403
            return view(**kwargs)
        return wrapped
    return decorator
